### ISSUE TITLE
Inline data for "image generator" inserted objects

### ISSUE THREAD

**Issue by jaap-karssenberg** [zim-desktop-wiki/zim-desktop-wiki#841]
Currently all "image generator" plugins (equation, diagram, ...) rely on a data file and an image file. This is not robust and contaminates the attachment folder.

Instead data should be inline in the wiki page like for other inserted object types - e.g. code blocks.

The image can than be stored in a cache file using a hash of the data as the file name. Similar to how thumbnails are stored.

This will also adres issue #418.

**Comment by pfalcon** [comment]
The benefit of the current solution is that any object is represented as just image tag (with a `type` query param). This is very smart and scalable scheme, e.g. any 3rd-party tool don't need to know anything about a particular object, it can just show a pre-rendered image instead.

Similarly, this offers an instant method to support arbitrary plugin for Markdown-format notebooks. In https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/26#issuecomment-300757202, you write:

> Probably also requires disabling some features in zim when source is markdown (e.g. insert of rich objects).

But there's absolutely no need to do that. In Markdown, any object can be just represented by native Markdown image syntax, just as in Zim-wiki format, objects are represented by native image syntax. I.e. what's now `{{./diagram.png?type=diagram}}` in Markdown would be `![diagram.png](./diagram.png?type=diagram)`.

Summing up, I'd recommend to be cautious about this change, weigh pros and cons. I guess I personally would be -1 on it due to reasons above.

**Comment by jaap-karssenberg** [comment]
I think for the native markdown format the ability to edit inline is more
important than rendering.

Probably there will still be an markdown export format flavor.

Op vr 11 sep. 2020 22:09 schreef Paul Sokolovsky <notifications@github.com>:

**Comment by pfalcon** [comment]
> I think for the native markdown format the ability to edit inline is more
important than rendering.

It seems that WYSIWYG (Rich Text) editing is "selling point" of Zim: how it differs from many other similar apps, and what users like and don't want to give up (based on the comments in https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/26).

Now suppose Rich Text *editing* is not available (e.g. on mobile with a 3rd-party Markdown viewer). Would you be ok with giving up Rich Text viewing too (I mean all the diagrams and other objects supported by plugins)? I definitely wouldn't be ok with that. Again, the idea is to pre-render such objects as just images, and reference them as such is very, very smart idea. And it's still very possible to edit data for objects - just not "inline", but switching to another file.

> Probably there will still be an markdown export format flavor.

Sure, it will have many knobs for user to adjust the output it produces. The talk here is what flavor of Markdown Zim should support as its native format, which flavor will allow to support more of Zim's "look and feel", without compromising on interoperability with 3rd-party applications.

For completeness, it's worth mentioning that it might be possible to support both "linking" and "embedding" cases. That would be the best of both worlds, but unclear what effect it would have on UX for non-advanced users.

And just for completeness, I'm writing this with awareness that there's Markdown extension syntax to embed limited types of objects ([random inet link](https://markdownmonster.west-wind.com/docs/_5ef0x96or.htm)). But it's clear, that support for such extension is very far from universal across Markdown applications, and set of supported object types is very limited comparing to Zim's support of arbitrary objects via plugins.

**Comment by jaap-karssenberg** [comment]
Framework added in 55ab6f2b69ee3e745bd267e56f11a53b7ad10aa3

Remaining works is to add plugins which use (and test) it, and let existing plugins (optionally ?) use it.
